nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+          idempotent manner.
+        * No other method should atler ``fields`` or ``query``.
+
+        :param results: A list of dicts, where each dict is a set of attributes
+            for one entity. The contents of these dicts are as is returned from
+            the server.
+        :returns: A list of dicts, where each dict is a set of attributes for
+            one entity. The contents of these dicts have been normalized and
+            can be used to instantiate entities.
+
+        """
+        fields = self.get_fields()
+        normalized = []
+        for result in results:
+            # For each field that we know about, copy the corresponding field
+            # from the server's search result. If any extra attributes are
+            # copied over, Entity.__init__ will raise a NoSuchFieldError.
+            # Examples of problematic results from server:
+            #
+            # * organization_id (denormalized OneToOne. see above)
+            # * organizations, organization_ids (denormalized OneToMany. above)
+            # * updated_at, created_at (these may be handled in the future)
+            # * sp_subnet (Host.sp_subnet is an undocumented field)
+            #
+            attrs = {}
+            for field_name, field in fields.items():
+                if isinstance(field, OneToOneField):
+                    try:
+                        attrs[field_name] = _get_entity_id(field_name, result)
+                    except MissingValueError:
+                        pass
+                elif isinstance(field, OneToManyField):
+                    try:
+                        attrs[field_name] = _get_entity_ids(field_name, result)
+                    except MissingValueError:
+                        pass
+                else:
+                    try:
+                        attrs[field_name] = result[field_name]
+                    except KeyError:
+                        pass
+            normalized.append(attrs)
+        return normalized
+
+    def search(self, fields=None, query=None, filters=None):
+        """Search for entities.
+
+        At its simplest, this method searches for all entities of a given kind.
+        For example, to ask for all :class:`nailgun.entities.Organization`
+        entities::
+
+            Organization().search()
+
+        Values on an entity are used to generate a search query, and the
+        ``fields`` argument can be used to specify which fields should be used
+        when generating a search query::
+
+            org = Organization(name='foo', description='bar')
+            results = org.search()  # Search by name and description.
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
```

Comment:
As of right now, all of these statements hold true:

```python
Organization().search() == Organization().search(fields=set())
Organization().search() == Organization().search(set())
Organization().search() == org.search(fields=set())
Organization().search() == org.search(set())
```

I've done this in order to make `search()` similar to `update()`. `update()` has this signature:

```python
def update(self, fields=None):
```

By default, `update()` sends all fields:

```python
org = entities.Organization(id=1, description='new description').update()
```

But you can also choose which fields to send:

```python
org = entities.Organization(id=1).read()
org.description = 'new description'
org = org.update(['description'])
```

Similarly, `search()` uses all fields:

```python
orgs = entities.Organization(description='new description').search()
```

But you may want to choose which fields to search by:

```python
org = entities.Organization(id=1).read()
orgs = org.search({'name'})
assert org.name == orgs[0].name
```

Now, you absolutely are right in pointing out that it's unusual to ask for a `set()` as an argument. That's pretty new.

Thoughts?